contrib/datadog/tracing/interceptor.go

```diff
@@ -0,0 +1,200 @@
+	_ interceptor.Tracer     = tracerImpl{}
+	_ interceptor.TracerSpan = &tracerSpan{}
+)
+
+// NewInterceptor creates an interceptor for setting on client options
+// that implements Datadog tracing for workflows.
+func NewInterceptor() interceptor.Tracer {
+	return tracerImpl{}
+}
+
+type contextKey string
+
+const (
+	activeSpanContextKey contextKey = "dd_trace_span"
+	headerKey                       = string(activeSpanContextKey)
+)
+
+type tracerImpl struct {
+	interceptor.BaseTracer
+}
+
+func (tracerImpl) Options() interceptor.TracerOptions {
+	return interceptor.TracerOptions{
+		SpanContextKey: activeSpanContextKey,
+		HeaderKey:      headerKey,
+	}
+}
+
+func (tracerImpl) UnmarshalSpan(m map[string]string) (interceptor.TracerSpanRef, error) {
+	var carrier tracer.TextMapCarrier = m
+	ctx, err := tracer.Extract(carrier)
+	if err != nil {
+		if errors.Is(err, tracer.ErrSpanContextNotFound) {
+			// If there is no span, return nothing, but don't error out. This is
+			// a legitimate place where a span does not exist in the headers
+			return nil, nil
+		}
+		return nil, err
+	}
+	return &tracerSpanCtx{ctx}, nil
+}
+
+func (tracerImpl) MarshalSpan(span interceptor.TracerSpan) (map[string]string, error) {
+	carrier := tracer.TextMapCarrier{}
+	if err := tracer.Inject(span.(*tracerSpan).Context(), carrier); err != nil {
+		return nil, err
+	}
+	return carrier, nil
+}
+
+func (tracerImpl) SpanFromContext(ctx context.Context) interceptor.TracerSpan {
+	span, ok := tracer.SpanFromContext(ctx)
+	if !ok {
+		return nil
+	}
+	return &tracerSpan{Span: span}
+}
+
+func (tracerImpl) ContextWithSpan(ctx context.Context, span interceptor.TracerSpan) context.Context {
+	return tracer.ContextWithSpan(ctx, span.(*tracerSpan).Span)
+}
+
+func (tracerImpl) StartSpan(options *interceptor.TracerStartSpanOptions) (interceptor.TracerSpan, error) {
+	startOpts := []tracer.StartSpanOption{
+		tracer.ResourceName(options.Name),
+		tracer.StartTime(options.Time),
+	}
+
+	// Set a deterministic span ID for workflows which are long-running and cross process boundaries
+	if options.Operation == "RunWorkflow" {
+		rid, ok := options.Tags["temporalRunID"]
+		if !ok {
+			return nil, fmt.Errorf("missing required tag `temporalRunID` on `RunWorkflow` span")
+		}
+		h := fnv.New64()
+		// Write() always writes all bytes and never fails; the count and error result are for implementing io.Writer.
+		_, _ = h.Write([]byte(fmt.Sprintf("workflow:%s", rid)))
+		startOpts = append(startOpts, tracer.WithSpanID(h.Sum64()))
+	}
+
+	// Add parent span to start options
+	var parent ddtrace.SpanContext
+	switch opParent := options.Parent.(type) {
+	case nil:
+	case *tracerSpan:
+		parent = opParent.Context()
+	case *tracerSpanCtx:
+		parent = opParent
+	default:
+		return nil, fmt.Errorf("unrecognized parent type %T", opParent)
+	}
+	if parent != nil {
+		// Convert the parent context into a full trace (required, otherwise
+		// the implementation of StartSpan() will NOT respect ChildOf()
```

Comment:
Nit: missing closing `)`. A new sentence avoids any parens.

```suggestion
		// Convert the parent context into a full trace.
		// This is required, otherwise the implementation of StartSpan()
		// will NOT respect ChildOf().
```